packages/flutter_tools/lib/src/buildconfig_generator.dart

```diff
@@ -0,0 +1,112 @@
+import 'globals.dart';
+
+const String _kFileName = 'build_config.g.dart';
+
+Future<Null> generateBuildConfigFile({@required BuildInfo buildInfo}) async {
+  return generateBuildConfigFileAtPath(buildInfo: buildInfo, path: '');
+}
+
+@visibleForTesting
+Future<Null> generateBuildConfigFileAtPath({@required BuildInfo buildInfo, @required String path}) async {
+  assert(buildInfo != null, 'buildInfo can\'t be null');
+  assert(path != null, 'path can\'t be null');
+
+  final Completer<Null> completer = new Completer<Null>();
+  final Status status = logger.startProgress('Generating $_kFileName file...', expectSlowOperation: true);
+
+  File buildConfig;
+  IOSink sink;
+
+  try {
+    buildConfig = fs.file(fs.path.join(path, 'lib', _kFileName));
+    if (buildConfig.existsSync()) {
+      buildConfig.deleteSync();
+    }
+    buildConfig.createSync();
+
+    sink = buildConfig.openWrite();
+    sink.write(_generateContent(buildInfo: buildInfo));
+    status.stop();
+    completer.complete();
+  } catch (error) {
+    printStatus('Failed to generate $_kFileName file');
+    printError('$error');
+    status.stop();
+    completer.completeError('Failed to change version properties. $error');
+  } finally {
+    await sink?.flush();
+    await sink?.close();
+  }
+
+  return completer.future;
+}
+
+String _generateContent({@required BuildInfo buildInfo}) {
+  assert(buildInfo != null, 'buildInfo can\'t be null');
+
+  final Class content = new Class((ClassBuilder b) {
+    b
+      ..name = 'BuildConfig'
+      ..docs.add('// GENERATED CODE - DO NOT MODIFY BY HAND\n')
+      ..fields.add(new Field((FieldBuilder b) => b
+        ..name = 'kBuildTime'
+        ..static = true
+        ..modifier = FieldModifier.constant
+        ..type = refer('String')
+        ..assignment = new Code('"${clock.now().toIso8601String()}"')))
+      ..fields.add(new Field((FieldBuilder b) => b
+        ..name = 'kBuildTimeMillisecondsSinceEpoch'
+        ..static = true
+        ..modifier = FieldModifier.constant
+        ..type = refer('int')
+        ..assignment = new Code('${clock.now().millisecondsSinceEpoch}')))
+      ..fields.add(new Field((FieldBuilder b) => b
+        ..name = 'kDebug'
+        ..static = true
+        ..modifier = FieldModifier.constant
+        ..type = refer('bool')
+        ..assignment = new Code('${buildInfo.isDebug}')))
+      ..fields.add(new Field((FieldBuilder b) => b
+        ..name = 'kModeName'
+        ..static = true
+        ..modifier = FieldModifier.constant
+        ..type = refer('String')
+        ..assignment = new Code('"${buildInfo.modeName}"')))
+      ..fields.add(new Field((FieldBuilder b) => b
+        ..name = 'kFlavor'
+        ..static = true
+        ..modifier = FieldModifier.constant
+        ..type = refer('String')
+        ..assignment = new Code('${buildInfo.flavor != null ? "buildInfo.flavor" : null }')))
+
+      // TODO wait for https://github.com/flutter/flutter/pull/16857 to be merged
```

Comment:
(minor nit: our syntax for TODOs is `// TODO(username): description, https://link_to_issue`